### PR TITLE
update package sources in repair

### PR DESCRIPTION
**Pull request by kevinushey**
Closes https://github.com/rstudio/renv/issues/1574.
Closes https://github.com/rstudio/renv/issues/1527.

### CODE REVIEW

File: tests/testthat/_snaps/repair.md [hunk 1]
```diff
@@ -0,0 +1,21 @@
+# repair() can update DESCRIPTION files for GitHub packages
+
+    Code
+      . <- repair()
+    Output
+      - No issues found with the project library's cache links.
```
Comment (kevinushey) [inline tests/testthat/_snaps/repair.md:6]: I sort of wonder if I should print little headers between each section.

File: NEWS.md [hunk 2]
```diff
@@ -1,7 +1,12 @@
 
 # renv 1.1.0  (UNRELEASED)
 
-* `status()` now first reports on uninstalled packages, before reporting on 
+* `renv::repair()` now checks for installed packages which lack an explicitly-declared
+  remote source, and updates their `DESCRIPTION` files if it can infer an appropriate
+  remote source. This typically occurs when a package is installed from local sources,
+  but appears to be maintained or published on a remote repository (e.g. GitHub). (#1574)
```
Comment (hadley) [inline NEWS.md:7]: Mention that this is no longer done in snapshot?

File: R/repair.R [hunk 3]
```diff
@@ -1,14 +1,21 @@
 
 #' Repair a project library
```
Comment (hadley) [inline R/repair.R:2]:
```suggestion
#' Repair project library and/or lockfile
```

File: R/repair.R [hunk 4]
```diff
@@ -1,14 +1,21 @@
...
 #'
-#' Repair a project library whose cache symlinks have become broken.
-#' renv will attempt to re-install the requisite packages.
+#' Use `repair()` to recover from some common issues that can occur with
+#' a project. Currently, two operations are performed:
+#'
+#' 1. Packages with broken symlinks into the cache will be re-installed,
```
Comment (hadley) [inline R/repair.R:7]:
```suggestion
#' 1. Packages with broken symlinks into the cache will be re-installed.
```

File: R/repair.R [hunk 5]
```diff
@@ -1,14 +1,21 @@
...
+#'
+#' 2. Packages which are installed from sources, but appear to be from
```
Comment (hadley) [inline R/repair.R:9]:
```suggestion
#' 2. Packages that were installed from sources, but appear to be from
```

File: tests/testthat/_snaps/repair.md (hunk 1 above)
Comment (hadley) [reply tests/testthat/_snaps/repair.md:6]: Yeah, maybe like:

```
# Library cache links
- No problems found
```

?

File: R/repair.R [hunk 6]
```diff
@@ -45,10 +60,97 @@ repair <- function(library  = NULL,
...
+  if (length(inferred) == 0L) {
+    writef("- All installed packages appear to be from a known source.")
+    return(TRUE)
+  }
+
+  # ask used
+  renv_scope_options(renv.verbose = TRUE)
+  renv_pretty_print(
+    c(
+      "The following package(s) do not have an explicitly-declared remote source.",
```
Comment (hadley) [inline R/repair.R:92]:
```suggestion
      "The following package(s) were installed from an unknown source.",
```
?

File: R/repair.R [hunk 7]
```diff
@@ -45,10 +60,97 @@ repair <- function(library  = NULL,
...
+      "However, renv was available to infer their remote sources from the DESCRIPTION file."
+    ),
+    sprintf("%s  [%s]", format(names(inferred)), inferred),
+    "renv::restore() may fail for packages without an explicitly-declared remote source."
+  )
+
+  choice <- menu(
+
+    choices =  c(
+      update = "Let renv to update the remote sources for these packages.",
```
Comment (hadley) [inline R/repair.R:102]:
```suggestion
      update = "Let renv infer the remote sources for these packages.",
```

File: R/repair.R [hunk 8]
```diff
@@ -45,10 +60,97 @@ repair <- function(library  = NULL,
...
+    writef("- All installed packages appear to be from a known source.")
+    return(TRUE)
+  }
+
+  # ask used
+  renv_scope_options(renv.verbose = TRUE)
+  renv_pretty_print(
+    c(
+      "The following package(s) do not have an explicitly-declared remote source.",
+      "However, renv was available to infer their remote sources from the DESCRIPTION file."
```
Comment (hadley) [inline R/repair.R:93]:
```suggestion
      "However, renv was available to infer the remote sources from their DESCRIPTION file."
```

File: R/repair.R [hunk 9]
```diff
@@ -45,10 +60,97 @@ repair <- function(library  = NULL,
...
+    choices =  c(
+      update = "Let renv to update the remote sources for these packages.",
+      cancel = "Do nothing and resolve the situation another way."
+    ),
+
+    title = "What would you like to do?"
+
+  )
+
+  if (identical(choice, "cancel")) {
```
Comment (hadley) [inline R/repair.R:110]: `cancel_if()`?

File: tests/testthat/_snaps/repair.md [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+      However, renv was available to infer their remote sources from the DESCRIPTION file.
+      - skeleton  [kevinushey/skeleton]
+      renv::restore() may fail for packages without an explicitly-declared remote source.
+      
+      What would you like to do?
+      
+      1: Let renv to update the remote sources for these packages.
+      2: Do nothing and resolve the situation another way.
+      
+      Selection: 2
```
Comment (hadley) [inline tests/testthat/_snaps/repair.md:18]: I wonder if we should eliminate the empty lines here now too?

**Review by hadley — APPROVED** [review]

**Pull request merged**